### ISSUE TITLE
'code' command doesn't work on remote when called from a GNU screen or tmux session.

### ISSUE THREAD

**Issue by nokeldin** [microsoft/vscode-remote-release#6362]
When opening a file in the editor using the CLI code command, it returns the following error:
```
Unable to connect to VS Code server: Error in request.
Error: connect ENOENT /run/user/46064/vscode-ipc-4037a58b-5b07-4250-bcbf-440659048f3b.sock
    at PipeConnectWrap.afterConnect [as oncomplete] (net.js:1146:16) {
  errno: -2,
  code: 'ENOENT',
  syscall: 'connect',
  address: '/run/user/46064/vscode-ipc-4037a58b-5b07-4250-bcbf-440659048f3b.sock'
}
```
IMPORTANT: This only happens if 'code' is called from inside a GNU screen/tmux session that was created from a different shell session than the one it is currently attached to. To clarify with an example:

(Steps to reproduce) 

1) Open a new remote terminal in VSCode ['code' command works fine]
2) Start a new screen/tmux session ['code' command works fine]
3) Detach from the sreen/tmux session ['code' command works fine]
4) Re-attach to the session ['code' command works fine]
5) Detach from the screen/tmux session and reattach from a new terminal --> 'code' stops working with the above error

- VSCode Version: 1.64.2
- Local OS Version: Windows 10 20H2
- Remote OS Version: CentOS 7
- Remote Extension/Connection Type: SSH/Docker/WSL: SSH

Does this issue occur when you try this locally?: No
Does this issue occur when you try this locally and all extensions are disabled?: No

**Comment by nokeldin** [comment]
As a workaround, I've been using the following bash function:
```
function fcode() {
  new_IPC_HOOK=$( lsof | grep $UID/vscode-ipc | awk '{print $NF}' | head -n 1 );
  echo $new_IPC_HOOK;
  export VSCODE_IPC_HOOK_CLI=$new_IPC_HOOK;
  if [[ ! -z $# ]]
  then
    code $@;
  fi
}
```
This function finds the earliest open VSCode ipc socket belonging to the current user. I haven't tested it on systems with multiple users running their VScode server instances, but should work regardless. Every time VSCode (client) is restarted, this function needs to be called once. If called without args, it will "fix" the current shell environment variables to point to the socket as found earlier. If called with args, it will fix the env var and then call "code" command with the specified args.

I don't know the implications of using the earliest open socket for this purpose, but it does work and hasn't caused any crashes or bugs yet.

It seems that VSCode will open one socket file every time a connection is established. This file will not be cleaned up at disconnect. Every time a new shell is started in the integrated terminal, another socket file is opened, but this one closes when the corresponding shell is killed. The script above uses the first socket (which doesn't seem to be in use by the shell sessions in the integrated terminal).

**Comment by posita** [comment]
See also #6346.

`socat` can be used to test and cleanup "dead" sockets [[source](https://unix.stackexchange.com/a/556790/106705)]:

``` sh
# Clean up the sockets that VS Code leaves all over the place like rabbit droppings. We
# use socat to test wether anyone is listening on the other side without reading or
# writing any bytes. If it’s a dead end, we assume the socket is stale and remove it.
for i in /tmp/vscode-*.sock ; do
    if ! socat -u OPEN:/dev/null "UNIX-CONNECT:${i}" ; then
        rm --force --verbose "${i}"
    fi
done
```

**Comment by max06** [comment]
It is intentional, for example for the ssh- and pgp-daemon. The client (gui) talks to the sockets on the client, passes the data to the code-server which itself creates sockets on the server to be used. Which absolutely makes sense and is (IMHO) a genius way to support multiple, different operating systems.

**Comment by posita** [comment]
My point was that VSCode could accommodate things like users using their own terminal emulators, `tmux`, etc. As your work-arounds have demonstrated, there's no technical reason why I shouldn't be able to `docker exec bash` into a container started by VSCode and use that to, e.g., `EDITOR='code --wait' git commit`.

Currently, the mechanisms that require that to work are:

1. Starting the shell from within VSCode (in which case, you are tied to VSCode's terminal emulator); or
2. For every customer of VSCode to independently field work-arounds nearly identical to yours. (Hint: That's a feature request.)

Those are not the only possible options. Another might be to package VSCode's environment bootstrapping independently, which could be installed, e.g., via `dnf install vscode-env-extensions`, `apt-get install vscode-env-extensions`, etc. (While a different animal, virtualization platforms like VMware take this approach. You _can_ run a guest OS without them, but to access higher level integration features, there are guest OS-specific plugins, extensions, etc. that you can install to enable such features, if desired.) VSCode is _already_ aware of os-specific configuration it needs to function inside the container. It _already_ installs os-specific extensions there. Decoupling those (e.g., making available VSCode environment initialization to arbitrary login shells) could open the door to the kind of composable flexibility being sought by the filer of this issue.

**Comment by max06** [comment]
As I tried to explain above: The vscode client (gui) and the code-server are separated on remote systems, like in WSL, ssh and docker-environments. VSCode has no idea about your terminal environment. If you do `docker exec bash` into a container, how should anything inside know what to do?

That's the exact reason why I wrote that workaround.

**Comment by posita** [comment]
I think we are mostly in agreement? (If not, I'm not sure where the contention is.) I think my point is that your work-arounds (or something that achieves their effect) should be available to end users as part of the remote extension and be well documented. Users shouldn't have to search GitHub issues and copy/paste snippets to solve these issues. VSCode should make such things accessible via its own tooling, which is what I think the original filer of this issue is seeking. I think that's a reasonable request, and I think it's achievable in a sane way.

**Comment by max06** [comment]
I'm confused. My bad probably.

I think I wrote in your issue, that I'd still like to see an extension to the `devcontainer`-command on the client operating system, which would basically connect to the target the same way I did, including setting the proper environment variables. Maybe something like `devcontainer run $path/to/project bash`, resulting in `docker exec ... bash -c "env=socket... $command`. 

As long as the vscode gui is the only official way to connect to any remote code server, a similar workaround will always be necessary for custom connections.
